packages/flutter_tools/lib/src/base/build.dart

```diff
@@ -0,0 +1,136 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:convert' show JSON;
+
+import 'package:crypto/crypto.dart' show md5;
+import 'package:meta/meta.dart' show required;
+
+import 'file_system.dart';
+
+/// A function that writes a set of outputs, given a set of input files.
+typedef Future<Null> BuildFunction(Set<File> inputs);
+
+class BuildException implements Exception {
+  BuildException(this.message, {this.cause}) {
+    assert(message != null);
+  }
+
+  final String message;
+  final dynamic cause;
+
+  @override
+  String toString() => message;
+}
+
+/// A build action that generates output artifacts given a set of inputs and a
+/// build function.
+///
+/// Build actions enforce the following rules:
+/// 1. There must be at least one input.
+/// 2. There must be at least one output.
+/// 3. No input may also be an output.
+/// 3. All declared inputs exist prior to the build.
+/// 4. All declared outputs exist after the build.
+///
+/// If a [checksumPath] is specified, a checksum is computed for every input
+/// file prior to the build, and recorded to the path specified. If all output
+/// files are present and the input file checksums are unchanged from the
+/// previous build, the build action is skipped, and the existing outputs are
+/// assumed valid.
+///
+/// As a consequence, if [checksumPath] is specified, all output files must be
+/// purely a function of the declared input files; otherwise, stale build
+/// results may be returned.
+class BuildAction {
+  BuildAction({
+    @required this.inputPaths,
+    @required this.outputPaths,
+    @required this.buildFunction,
+    String checksumPath,
+  }) : _checksumPath = checksumPath {
+    assert(inputPaths != null && inputPaths.isNotEmpty);
```

Comment:
By default, we don't run the Flutter tool in checked mode.  You may want to turn these into non-checked-mode checks.